<a href="https://colab.research.google.com/github/john-decker/john-decker-Arts-and-Humanities-Programming-CoLab-Work/blob/main/Lecture6_Data_Wrangling_Part2_Cleaning_and_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###Using Pandas
The Pandas library was created to enable complex data analysis. Currently, it is the go-to tool for Data Scientists thanks to its speed and flexibility. Pandas is useful for cleaning data, exploring data, analyzing data, outputting data to other applications (such as machine learning), and visualizing data. The library is built on the NumPy library, which is written in a combination of Python and C++. NumPy is optimized for fast calculation of sequential data, which makes it well-suited to scientific/mathematical tasks. 

Pandas allows us to work efficiently with series data (one-dimensional array data) as well as data frame data (two dimensional array data). The library has built-in functionality that allows us to shape, group, subset, and merge data. We can also work with the data using high-level math packages and can carry out various types of statistical analysis with tools built-in to Pandas. 

Because we can work with our data programmatically, we have a great deal of flexibility in how we move through various stages: initial exploration, cleaning and organization, analysis, modeling, and output (graphs, tables, etc.). The scripts that we create to produce results, can be shared with others who can apply them to the same data and reproduce our findings, which helps validate the claims we make about the data we are analyzing.


In [ ]:
#read in our data and preview its contents
import pandas as pd

big_data_filepath = '/content/2015_Street_Tree_Census_Tree_Data_First_2K_cleaned.csv'

trees_df = pd.read_csv(big_data_filepath)

trees_df.head


<bound method NDFrame.head of       tree_id  block_id  created_at  tree_dbh  stump_diam curb_loc status  \
0      180683    348711  08/27/2015         3           0   OnCurb  Alive   
1      200540    315986  09/03/2015        21           0   OnCurb  Alive   
2      204026    218365  09/05/2015         3           0   OnCurb  Alive   
3      204337    217969  09/05/2015        10           0   OnCurb  Alive   
4      189565    223043  08/30/2015        21           0   OnCurb  Alive   
...       ...       ...         ...       ...         ...      ...    ...   
1909   203622    348207  09/04/2015         8           0   OnCurb  Alive   
1910   209089    503966  09/08/2015         5           0   OnCurb  Alive   
1911   179390    305893  08/27/2015         5           0   OnCurb  Alive   
1912   198157    411913  09/02/2015         8           0   OnCurb  Alive   
1913   209323    414316  09/08/2015         5           0   OnCurb  Alive   

     health                           spc_lat

In [ ]:
#get the 'shape' of our data (number of rows and columns)
trees_df.shape

(1914, 48)

In [ ]:
#get the names of the variables 
trees_df.columns

Index(['tree_id', 'block_id', 'created_at', 'tree_dbh', 'stump_diam',
       'curb_loc', 'status', 'health', 'spc_latin', 'spc_common', 'steward',
       'guards', 'sidewalk', 'user_type', 'problems 1', 'problems 2',
       'problems 3', 'problems 4', 'root_stone', 'root_grate', 'root_other',
       'trunk_wire', 'trnk_light', 'trnk_other', 'brch_light', 'brch_shoe',
       'brch_other', 'address', 'postcode', 'zip_city', 'community board',
       'borocode', 'borough', 'cncldist', 'st_assem', 'st_senate', 'nta',
       'nta_name', 'boro_ct', 'state', 'latitude', 'longitude', 'x_sp', 'y_sp',
       'council district', 'census tract', 'bin', 'bbl'],
      dtype='object')

In [ ]:
#get series data from a particular column
trees_df.spc_common

0             red maple
1               pin oak
2           honeylocust
3           honeylocust
4       American linden
             ...       
1909       Callery pear
1910       Callery pear
1911       Callery pear
1912       Callery pear
1913       Callery pear
Name: spc_common, Length: 1914, dtype: object

In [ ]:
#iterate over series to get all values
tree_names_english = trees_df.spc_common
for tree in tree_names_english.iteritems():
    print(tree)

(0, 'red maple')
(1, 'pin oak')
(2, 'honeylocust')
(3, 'honeylocust')
(4, 'American linden')
(5, 'honeylocust')
(6, 'honeylocust')
(7, 'American linden')
(8, 'honeylocust')
(9, 'London planetree')
(10, 'London planetree')
(11, 'London planetree')
(12, 'London planetree')
(13, 'London planetree')
(14, 'London planetree')
(15, 'honeylocust')
(16, 'ginkgo')
(17, 'honeylocust')
(18, 'honeylocust')
(19, 'ginkgo')
(20, 'honeylocust')
(21, 'honeylocust')
(22, 'willow oak')
(23, 'sycamore maple')
(24, 'willow oak')
(25, 'Amur maple')
(26, 'London planetree')
(27, 'London planetree')
(28, 'honeylocust')
(29, 'honeylocust')
(30, 'honeylocust')
(31, 'honeylocust')
(32, 'honeylocust')
(33, 'honeylocust')
(34, 'honeylocust')
(35, 'honeylocust')
(36, 'honeylocust')
(37, 'London planetree')
(38, 'red maple')
(39, 'London planetree')
(40, 'London planetree')
(41, 'willow oak')
(42, 'London planetree')
(43, 'London planetree')
(44, 'London planetree')
(45, 'willow oak')
(46, 'London planetree')
(47, 'L

In [ ]:
#access specific values from the tuple returned by iteration
tree_names_english = trees_df.spc_common
for tree in tree_names_english.iteritems():
    print(tree[1])

red maple
pin oak
honeylocust
honeylocust
American linden
honeylocust
honeylocust
American linden
honeylocust
London planetree
London planetree
London planetree
London planetree
London planetree
London planetree
honeylocust
ginkgo
honeylocust
honeylocust
ginkgo
honeylocust
honeylocust
willow oak
sycamore maple
willow oak
Amur maple
London planetree
London planetree
honeylocust
honeylocust
honeylocust
honeylocust
honeylocust
honeylocust
honeylocust
honeylocust
honeylocust
London planetree
red maple
London planetree
London planetree
willow oak
London planetree
London planetree
London planetree
willow oak
London planetree
London planetree
London planetree
willow oak
willow oak
London planetree
London planetree
London planetree
London planetree
pin oak
pin oak
pin oak
pin oak
red maple
red maple
pin oak
pin oak
hedge maple
pin oak
pin oak
pin oak
London planetree
London planetree
London planetree
London planetree
American Elm
sycamore maple
London planetree
London planetree
London planetre

In [ ]:
#get counts by species in ascending order
tree_counts_ascending = trees_df.spc_common.value_counts(ascending=True)
tree_counts_ascending

katsura tree              1
spruce                    1
crepe myrtle              1
pignut hickory            1
Ohio buckeye              1
weeping willow            1
blackgum                  1
southern magnolia         1
Turkish hazelnut          1
English oak               1
Chinese fringetree        1
Schumard's oak            1
white ash                 1
ash                       2
sassafras                 2
Douglas-fir               2
mulberry                  2
Kentucky yellowwood       2
Atlantic white cedar      2
holly                     2
white pine                2
Chinese elm               3
black walnut              3
cherry                    3
Siberian elm              3
golden raintree           3
sycamore maple            4
black cherry              4
horse chestnut            5
hedge maple               5
hardy rubber tree         7
eastern redcedar          7
black oak                 8
scarlet oak               8
sawtooth oak              8
Amur maple          

In [ ]:
#get counts by species in descending order
tree_counts_descending = trees_df.spc_common.value_counts(ascending=False)
print(tree_counts_descending)


honeylocust             323
Norway maple            308
London planetree        249
pin oak                 163
ginkgo                  116
Sophora                 109
Callery pear            106
Japanese zelkova         69
American linden          51
littleleaf linden        49
silver maple             30
sweetgum                 30
Japanese tree lilac      29
red maple                29
swamp white oak          28
American Elm             20
green ash                18
northern red oak         16
tulip-poplar             14
willow oak               13
white oak                12
crab apple               11
silver linden            11
Amur maple               10
black oak                 8
sawtooth oak              8
scarlet oak               8
hardy rubber tree         7
eastern redcedar          7
horse chestnut            5
hedge maple               5
sycamore maple            4
black cherry              4
Chinese elm               3
black walnut              3
Siberian elm        

In [ ]:
#get series data on the diameter of the trees in the data set
#use .describe() to get basic statistics on the data
tree_diam = trees_df.tree_dbh
print(tree_diam.describe())

count    1914.000000
mean       11.668757
std         7.832672
min         1.000000
25%         5.000000
50%        10.000000
75%        16.000000
max        44.000000
Name: tree_dbh, dtype: float64


From the data dictionary on the NYC Open Data portal, we see that “tree_dbh” is a measure of the trunk diameter at 54” (or 137 cm) above the ground. 

**Note** that although the data dictionary gives the metric value for the height, the actual data appear to be solely measured in inches. 

**Remember** that our data is a non-randomly sampled subset of the much larger dataset. As a result, we have to be careful with things like mean and standard deviation. These values could be wildly different with the inclusion of the rest of the data. What we can say is that for the sample of 1,914 trees in our data set, these are the mean and standard deviation numbers. 

The same holds true for the percentiles. That said, the data in our set show that the 50th percentile (or the number under which 50% of the data points fall) is 10 inches. 

**Note** that this is smaller than the mean, which indicates that the data are skewed. In this case, they are “right skewed” meaning that there is a “long tail” on the right of the graph. 

In a “normal” distribution, the mean, median, and mode (and hence the 50th percentile point) are all the same and fall in the middle of the data. The data above and below that point are evenly distributed and form a “bell curve”. 

In a skewed dataset, like the one we have here, the data are not evenly distributed around the center. Our tree data is more heavily clustered among the smaller trunk diameters. The presence of much larger trunks (e.g. 44”) create ‘outliers’ that pull the “tail” of the data toward the right. 

We can get a sense of just how far away from the other numbers the maximum is by looking at the 75th percentile. In our data, 75% of the tree diameters fall below 16”. As we can see, 44” is a good distance from the majority of the measurements in the series. Pandas allows us to measure the amount of skew we have.

In [ ]:
tree_diam.skew()

1.0430949278044486

**Be aware** that skewed data does not automatically mean “bad” data. In natural systems (like the trunks of trees) normal variation can lead to skewing. The question then becomes what are potential reasons for the skew? It could be our non-random sampling. It could be that some tree species grow better or worse in the urban environment of NYC. It might also be that there are more tree species in our sample that are prone to being smaller–the choice to plant them along with bigger species could also drive the variation. It could be a combination of one or more of these factors, or something else entirely. In other words, we require more context to understand our data. This is critical – **data do not “mean” anything by themselves**. We have to put them into context before the numbers mean anything useful at all.

###Context for Data
Let’s return to our count of trees to see which were the most numerous.

```print(tree_counts_descending)```

|Species |Count |
|:---    |:---  |
|honeylocust |           323|
Norway maple |           308|
London planetree |       249|
pin oak  |                163|
ginkgo   |              116|
Sophora  |             109|
Callery pear |           106|

We’ll focus on species with more than 100 instances in the data–it’s as good a place as any to begin as these species account for 1,374 of our 1,914 data points (i.e. approx. 72%). 

At the bottom is the Callery pear. If we do a tiny bit of research on this species, we see that the tree tends to be on the small side. The [Penn State Extension](https://extension.psu.edu/callery-pear) website notes that the species rarely gets taller than 40 feet and its trunk reaches a maximum size of less than a foot in diameter (~36” diameter when full grown). The [Arbor Day Foundation](https://www.arborday.org/trees/treeguide/TreeDetail.cfm?Itemid=2578) site notes that the tree is fast-growing, putting on an average of 24” of height per year. We also discover that it is a problematic tree – it is considered an invasive species and is often put on a list of trees that need to be “controlled”. The [North Carolina State Extension](https://plants.ces.ncsu.edu/plants/pyrus-calleryana/) website echoes this by stating that: 

Limb breakage is common due to poor crotch angles and it is often a very short-lived tree.  This is a weedy to invasive species, that can spread fireblight to other Rosaceous crops, so consider planting a different tree.

In addition, it is short-lived (only 10 to 20 years) and produces an offensive odor, which is likened to rotting animals, in the Spring. So, why plant it at all? It is known for its ornamental flowers and it is suited to places like NYC. An [information sheet](https://www.oakopenings.org/wp-content/uploads/2019/09/BMP_CalleryPear_finalversion.pdf) about the species prepared for Northwest Ohio and Southeast Michigan notes that the tree is popular in cities because of its resistance to urban stresses (poor soil, dry conditions, heat, and pollution).

So, how does this information provide any context for our dataset? Out of all the trees in our dataset, the Callery Pear accounts for around 5.5%. Its maximum trunk diameter is above the 75th percentile. These two factors may indicate that this species is part of the long right tail of the data causing some of our skewing. More importantly, it points to NYC making planting decisions based more on visual aesthetics than longevity. It also shows that NYC is willing to harbor an invasive species because it can tolerate the inhospitable conditions of the city itself. What kinds of consequences might these decisions have on issues of sustainability and of environmental protection?

**Now it’s your turn:** choose one of the tree species in the list and find out more about it. Share your findings with your classmates on the course Canvas site.

In [ ]:
#slicing dataframes using iloc()
trees_df.iloc[:12]

tree_id  block_id  created_at  tree_dbh  stump_diam        curb_loc  \
0    180683    348711  08/27/2015         3           0          OnCurb   
1    200540    315986  09/03/2015        21           0          OnCurb   
2    204026    218365  09/05/2015         3           0          OnCurb   
3    204337    217969  09/05/2015        10           0          OnCurb   
4    189565    223043  08/30/2015        21           0          OnCurb   
5    190422    106099  08/30/2015        11           0          OnCurb   
6    190426    106099  08/30/2015        11           0          OnCurb   
7    208649    103940  09/07/2015         9           0          OnCurb   
8    209610    407443  09/08/2015         6           0          OnCurb   
9    192755    207508  08/31/2015        21           0  OffsetFromCurb   
10   203719    302371  09/05/2015        11           0          OnCurb   
11   203726    302371  09/05/2015         8           0          OnCurb   

   status health                           spc_latin        spc_common  ...  \
0   Alive   Fair                         Acer rubrum         red maple  ...   
1   Alive   Fair                   Quercus palustris           pin oak  ...   
2   Alive   Good  Gleditsia triacanthos var. inermis       honeylocust  ...   
3   Alive   Good  Gleditsia triacanthos var. inermis       honeylocust  ...   
4   Alive   Good                     Tilia americana   American linden  ...   
5   Alive   Good  Gleditsia triacanthos var. inermis       honeylocust  ...   
6   Alive   Good  Gleditsia triacanthos var. inermis       honeylocust  ...   
7   Alive   Good                     Tilia americana   American linden  ...   
8   Alive   Good  Gleditsia triacanthos var. inermis       honeylocust  ...   
9   Alive   Fair               Platanus x acerifolia  London planetree  ...   
10  Alive   Good               Platanus x acerifolia  London planetree  ...   
11  Alive   Poor               Platanus x acerifolia  London planetree  ...   

    boro_ct     state   latitude  longitude          x_sp         y_sp  \
0   4073900  New York  40.723092 -73.844215  1.027431e+06  202756.7687   
1   4097300  New York  40.794111 -73.818679  1.034456e+06  228644.8374   
2   3044900  New York  40.717581 -73.936608  1.001823e+06  200716.8913   
3   3044900  New York  40.713537 -73.934456  1.002420e+06  199244.2531   
4   3016500  New York  40.666778 -73.975979  9.909138e+05  182202.4260   
5   1014500  New York  40.770046 -73.984950  9.884187e+05  219825.5227   
6   1014500  New York  40.770210 -73.985338  9.883112e+05  219885.2785   
7   1012700  New York  40.762724 -73.987297  9.877691e+05  217157.8561   
8   5006400  New York  40.596579 -74.076255  9.630732e+05  156635.5542   
9   3037402  New York  40.586357 -73.969744  9.926537e+05  152903.6306   
10  4010500  New York  40.782428 -73.911171  1.008850e+06  224349.0366   
11  4010500  New York  40.781735 -73.912020  1.008615e+06  224096.2740   

   council district census tract        bin           bbl  
0              29.0        739.0  4052307.0  4.022210e+09  
1              19.0        973.0  4101931.0  4.044750e+09  
2              34.0        449.0  3338310.0  3.028870e+09  
3              34.0        449.0  3338342.0  3.029250e+09  
4              39.0        165.0  3025654.0  3.010850e+09  
5               3.0        145.0  1076229.0  1.011310e+09  
6               3.0        145.0  1076229.0  1.011310e+09  
7               3.0        133.0  1086093.0  1.010410e+09  
8               NaN          NaN        NaN           NaN  
9              47.0      37402.0  3320727.0  3.072350e+09  
10             22.0        105.0  4019061.0  4.008710e+09  
11             22.0        105.0  4019059.0  4.008710e+09  

[12 rows x 48 columns]

In [ ]:
#slicing using [start:stop:step] syntax
trees_df.iloc[4:30:3]

tree_id  block_id  created_at  tree_dbh  stump_diam curb_loc status  \
4    189565    223043  08/30/2015        21           0   OnCurb  Alive   
7    208649    103940  09/07/2015         9           0   OnCurb  Alive   
10   203719    302371  09/05/2015        11           0   OnCurb  Alive   
13   189465    219493  08/30/2015        22           0   OnCurb  Alive   
16   204208    204913  09/05/2015         2           0   OnCurb  Alive   
19   208201    219258  09/07/2015        10           0   OnCurb  Alive   
22   199760    107038  09/03/2015         4           0   OnCurb  Alive   
25   205044    206800  09/05/2015         5           0   OnCurb  Alive   
28   201555    102919  09/04/2015         9           0   OnCurb  Alive   

   health                           spc_latin        spc_common  ...  boro_ct  \
4    Good                     Tilia americana   American linden  ...  3016500   
7    Good                     Tilia americana   American linden  ...  1012700   
10   Good               Platanus x acerifolia  London planetree  ...  4010500   
13   Good               Platanus x acerifolia  London planetree  ...  3019100   
16   Fair                       Ginkgo biloba            ginkgo  ...  3030400   
19   Good                       Ginkgo biloba            ginkgo  ...  3018300   
22   Fair                     Quercus phellos        willow oak  ...  1016300   
25   Good                        Acer ginnala        Amur maple  ...  3030200   
28   Good  Gleditsia triacanthos var. inermis       honeylocust  ...  1003601   

       state   latitude  longitude          x_sp         y_sp  \
4   New York  40.666778 -73.975979  9.909138e+05  182202.4260   
7   New York  40.762724 -73.987297  9.877691e+05  217157.8561   
10  New York  40.782428 -73.911171  1.008850e+06  224349.0366   
13  New York  40.694733 -73.968211  9.930653e+05  192388.0651   
16  New York  40.593788 -73.991597  9.865838e+05  155609.3384   
19  New York  40.691499 -73.972588  9.918519e+05  191209.4250   
22  New York  40.782087 -73.980964  9.895217e+05  224212.8876   
25  New York  40.594936 -73.989171  9.872573e+05  156027.9255   
28  New York  40.721807 -73.989830  9.870691e+05  202250.6142   

   council district census tract        bin           bbl  
4              39.0        165.0  3025654.0  3.010850e+09  
7               3.0        133.0  1086093.0  1.010410e+09  
10             22.0        105.0  4019061.0  4.008710e+09  
13             35.0        191.0  3054331.0  3.018880e+09  
16             47.0        304.0  3186635.0  3.068940e+09  
19             35.0        183.0  3058447.0  3.020890e+09  
22              6.0        163.0  1030859.0  1.011680e+09  
25             47.0        302.0  3186229.0  3.068790e+09  
28              1.0       3601.0  1077637.0  1.004160e+09  

[9 rows x 48 columns]

In [ ]:
#"filtering" by checking for a condition
# Note: if you do not enclose in print(), CoLab will print the info in spreadsheet format
just_pears = trees_df.loc[trees_df['spc_common'] == 'Callery pear']
just_pears

tree_id  block_id  created_at  tree_dbh  stump_diam curb_loc status  \
409    188408    340716  08/30/2015        10           0   OnCurb  Alive   
410    165046    412972  08/22/2015         6           0   OnCurb  Alive   
411    194919    227482  09/01/2015        12           0   OnCurb  Alive   
412    200201    406819  09/03/2015        11           0   OnCurb  Alive   
413    200210    406819  09/03/2015         8           0   OnCurb  Alive   
...       ...       ...         ...       ...         ...      ...    ...   
1909   203622    348207  09/04/2015         8           0   OnCurb  Alive   
1910   209089    503966  09/08/2015         5           0   OnCurb  Alive   
1911   179390    305893  08/27/2015         5           0   OnCurb  Alive   
1912   198157    411913  09/02/2015         8           0   OnCurb  Alive   
1913   209323    414316  09/08/2015         5           0   OnCurb  Alive   

     health         spc_latin    spc_common  ...  boro_ct     state  \
409    Good  Pyrus calleryana  Callery pear  ...  4028700  New York   
410    Good  Pyrus calleryana  Callery pear  ...  5017008  New York   
411    Good  Pyrus calleryana  Callery pear  ...  3044000  New York   
412    Good  Pyrus calleryana  Callery pear  ...  5029102  New York   
413    Good  Pyrus calleryana  Callery pear  ...  5029102  New York   
...     ...               ...           ...  ...      ...       ...   
1909   Fair  Pyrus calleryana  Callery pear  ...  4070700  New York   
1910   Good  Pyrus calleryana  Callery pear  ...  2021700  New York   
1911   Poor  Pyrus calleryana  Callery pear  ...  4127700  New York   
1912   Good  Pyrus calleryana  Callery pear  ...  5017009  New York   
1913   Good  Pyrus calleryana  Callery pear  ...  5012806  New York   

       latitude  longitude          x_sp         y_sp council district  \
409   40.749121 -73.889623  1.014833e+06  212221.0471             25.0   
410   40.550177 -74.184095  9.330892e+05  139774.5558             51.0   
411   40.612277 -73.971444  9.921784e+05  162346.5438             44.0   
412   40.588682 -74.189132  9.317198e+05  153805.7331             50.0   
413   40.588047 -74.188113  9.320024e+05  153573.8707             50.0   
...         ...        ...           ...          ...              ...   
1909  40.715812 -73.852919  1.025023e+06  200100.4210             29.0   
1910  40.846719 -73.915331  1.007675e+06  247771.3831             14.0   
1911  40.725687 -73.787903  1.043038e+06  203734.8820             24.0   
1912  40.541246 -74.187474  9.321432e+05  136522.9252             51.0   
1913  40.559123 -74.113154  9.528081e+05  143000.2889             50.0   

     census tract        bin           bbl  
409         287.0  4029863.0  4.012870e+09  
410       17008.0  5077364.0  5.057120e+09  
411         440.0  3173681.0  3.065850e+09  
412       29102.0  5135160.0  5.026430e+09  
413       29102.0  5135176.0  5.026430e+09  
...           ...        ...           ...  
1909        707.0  4076285.0  4.031980e+09  
1910      21502.0  2008472.0  2.028660e+09  
1911       1277.0  4155514.0  4.072300e+09  
1912      17009.0  5080941.0  5.062860e+09  
1913      12806.0  5130898.0  5.041300e+09  

[106 rows x 48 columns]

In [ ]:
#connecting our condition to the stats we generated earlier
just_large_diameters = trees_df.loc[trees_df['tree_dbh'] >= 30]
just_large_diameters

tree_id  block_id  created_at  tree_dbh  stump_diam        curb_loc  \
14     192998    211160  08/31/2015        30           0          OnCurb   
26     192942    310298  08/31/2015        33           0          OnCurb   
43     193690    227528  09/01/2015        36           0          OnCurb   
53     192888    210568  08/31/2015        32           0          OnCurb   
73     196080    310272  09/02/2015        36           0          OnCurb   
...       ...       ...         ...       ...         ...             ...   
1713   196487    210467  09/02/2015        34           0  OffsetFromCurb   
1763   196485    210467  09/02/2015        34           0  OffsetFromCurb   
1826   199918    229847  09/03/2015        33           0          OnCurb   
1849   209003    301774  09/08/2015        32           0          OnCurb   
1886   181372    321910  08/28/2015        36           0  OffsetFromCurb   

     status health                spc_latin        spc_common  ...  boro_ct  \
14    Alive   Fair    Platanus x acerifolia  London planetree  ...  3090000   
26    Alive   Good    Platanus x acerifolia  London planetree  ...  4013800   
43    Alive   Good    Platanus x acerifolia  London planetree  ...  3044000   
53    Alive   Fair    Platanus x acerifolia  London planetree  ...  3054800   
73    Alive   Good    Platanus x acerifolia  London planetree  ...  4014000   
...     ...    ...                      ...               ...  ...      ...   
1713  Alive   Good        Quercus palustris           pin oak  ...  3066600   
1763  Alive   Good        Quercus palustris           pin oak  ...  3066600   
1826  Alive   Good    Platanus x acerifolia  London planetree  ...  3076800   
1849  Alive   Good        Quercus palustris           pin oak  ...  4012500   
1886  Alive   Good  Liquidambar styraciflua          sweetgum  ...  4113900   

         state   latitude  longitude          x_sp         y_sp  \
14    New York  40.664317 -73.921130  1.006131e+06  181314.9855   
26    New York  40.706534 -73.824992  1.032772e+06  196734.1647   
43    New York  40.613663 -73.974530  9.913215e+05  162851.2413   
53    New York  40.610035 -73.948634  9.985120e+05  161532.6868   
73    New York  40.704945 -73.828195  1.031885e+06  196153.4808   
...        ...        ...        ...           ...          ...   
1713  New York  40.603972 -73.936076  1.002000e+06  159326.2306   
1763  New York  40.603782 -73.935864  1.002059e+06  159257.0713   
1826  New York  40.621368 -73.958671  9.957232e+05  165660.2363   
1849  New York  40.772619 -73.918027  1.006955e+06  220773.2032   
1886  New York  40.762100 -73.788019  1.042974e+06  217001.4788   

     council district census tract        bin           bbl  
14               41.0        900.0  3081177.0  3.035310e+09  
26               29.0        138.0  4193186.0  4.092480e+09  
43               44.0        440.0  3173500.0  3.065820e+09  
53               48.0        548.0  3182700.0  3.067870e+09  
73               29.0        140.0  4193005.0  4.092390e+09  
...               ...          ...        ...           ...  
1713             46.0        638.0  3248115.0  3.088160e+09  
1763             46.0        638.0  3248149.0  3.088170e+09  
1826             44.0        768.0  3180489.0  3.067280e+09  
1849             22.0        125.0  4017460.0  4.008410e+09  
1886             19.0       1139.0  4120158.0  4.053200e+09  

[66 rows x 48 columns]

####Try it yourself:
Choose a range of trees you’d like to isolate and create a condition that will find only those trees. What happens if you use the “postcode” column and use your zipcode as the condition?


In [ ]:
my_zip_trees = trees_df.loc[trees_df['postcode'] == 11205]
my_zip_trees

tree_id  block_id  created_at  tree_dbh  stump_diam        curb_loc  \
13     189465    219493  08/30/2015        22           0          OnCurb   
15     189834    219505  08/30/2015        12           0          OnCurb   
19     208201    219258  09/07/2015        10           0          OnCurb   
47     189339    219520  08/30/2015        18           0  OffsetFromCurb   
83     208832    220174  09/08/2015         3           0          OnCurb   
...       ...       ...         ...       ...         ...             ...   
1855   189833    219505  08/30/2015         5           0          OnCurb   
1871   208868    219320  09/08/2015         3           0          OnCurb   
1897   189661    219530  08/30/2015         4           0          OnCurb   
1905   208185    221309  09/07/2015        11           0          OnCurb   
1908   189835    219505  08/30/2015        21           0          OnCurb   

     status health                           spc_latin        spc_common  ...  \
13    Alive   Good               Platanus x acerifolia  London planetree  ...   
15    Alive   Good  Gleditsia triacanthos var. inermis       honeylocust  ...   
19    Alive   Good                       Ginkgo biloba            ginkgo  ...   
47    Alive   Good               Platanus x acerifolia  London planetree  ...   
83    Alive   Poor               Platanus x acerifolia  London planetree  ...   
...     ...    ...                                 ...               ...  ...   
1855  Alive   Fair                               Morus          mulberry  ...   
1871  Alive   Good               Platanus x acerifolia  London planetree  ...   
1897  Alive   Good                   Quercus palustris           pin oak  ...   
1905  Alive   Good                    Pyrus calleryana      Callery pear  ...   
1908  Alive   Good                    Pyrus calleryana      Callery pear  ...   

      boro_ct     state   latitude  longitude         x_sp         y_sp  \
13    3019100  New York  40.694733 -73.968211  993065.3039  192388.0651   
15    3019100  New York  40.693314 -73.967601  993234.6165  191871.1586   
19    3018300  New York  40.691499 -73.972588  991851.9185  191209.4250   
47    3019100  New York  40.695281 -73.966380  993572.9187  192587.7409   
83    3019300  New York  40.692278 -73.961423  994947.9979  191494.4858   
...       ...       ...        ...        ...          ...          ...   
1855  3019100  New York  40.693325 -73.967512  993259.3271  191874.9506   
1871  3019300  New York  40.692148 -73.960751  995134.1593  191447.0515   
1897  3019100  New York  40.693530 -73.965742  993749.9072  191949.9052   
1905  3019300  New York  40.693406 -73.960889  995095.9091  191905.5431   
1908  3019100  New York  40.693304 -73.967691  993209.5106  191867.3059   

     council district census tract        bin           bbl  
13               35.0        191.0  3054331.0  3.018880e+09  
15               35.0        191.0  3054408.0  3.018890e+09  
19               35.0        183.0  3058447.0  3.020890e+09  
47               35.0        191.0  3054489.0  3.018910e+09  
83               35.0        193.0  3054898.0  3.019090e+09  
...               ...          ...        ...           ...  
1855             35.0        191.0  3054408.0  3.018890e+09  
1871             33.0        193.0  3343740.0  3.019248e+09  
1897             35.0        191.0  3054526.0  3.018910e+09  
1905             35.0        193.0  3399720.0  3.019098e+09  
1908             35.0        195.0        NaN           NaN  

[64 rows x 48 columns]

In addition to isolating entire rows, Pandas also allows us to group rows by the values in their columns. For example, what if we wanted to know which trees were in each region of NYC? We could use the .groupby() method to organize them by “zip_city” (i.e. Richmond Hill, Flushing, etc.) and the common name of the tree.

In [ ]:
trees_grouped = trees_df.groupby('zip_city')['spc_common']

for zip_code_pattern in trees_grouped:
    print(zip_code_pattern)

('Arverne', 395        sweetgum
396        sweetgum
456         Sophora
520         Sophora
1351        pin oak
1673    honeylocust
Name: spc_common, dtype: object)
('Astoria', 10         London planetree
11         London planetree
46         London planetree
74         London planetree
82         London planetree
               ...         
1875                 cherry
1884    Japanese tree lilac
1889               sweetgum
1895              red maple
1901           silver maple
Name: spc_common, Length: 94, dtype: object)
('Bayside', 301          silver maple
358          Norway maple
765              sweetgum
1171    littleleaf linden
1820       weeping willow
Name: spc_common, dtype: object)
('Bellerose', 1527    hardy rubber tree
Name: spc_common, dtype: object)
('Bronx', 31              honeylocust
38                red maple
57                  pin oak
69         London planetree
71             American Elm
               ...         
1846                pin oak
1873       Londo

**Be aware** that .groupby() is a method associated with series data – sometimes when working with dataframes, you get a series object as a result. The results of our grouping can be a bit hard to read, so let’s isolate things a bit. Let’s find a way to get the information in a form where we have the zip city followed by a list of the trees associated with it.

In [ ]:
for item in trees_grouped:
    region = item[0]
    trees = item[1].tolist()
    print(f'{region}:{trees}')

Arverne:['sweetgum', 'sweetgum', 'Sophora', 'Sophora', 'pin oak', 'honeylocust']
Astoria:['London planetree', 'London planetree', 'London planetree', 'London planetree', 'London planetree', 'London planetree', 'London planetree', 'London planetree', 'pin oak', 'honeylocust', 'honeylocust', 'London planetree', 'London planetree', 'London planetree', 'Norway maple', 'London planetree', 'Callery pear', 'American linden', 'honeylocust', 'Sophora', 'honeylocust', 'pin oak', 'Norway maple', 'Norway maple', 'Norway maple', 'Norway maple', 'Norway maple', 'Norway maple', 'Norway maple', 'Norway maple', 'Norway maple', 'honeylocust', 'tulip-poplar', 'pin oak', 'tulip-poplar', 'honeylocust', 'honeylocust', 'Norway maple', 'Norway maple', 'Norway maple', 'Norway maple', 'Norway maple', 'Norway maple', 'Norway maple', 'honeylocust', 'honeylocust', 'London planetree', 'Japanese zelkova', 'black oak', 'Norway maple', 'London planetree', 'littleleaf linden', 'littleleaf linden', 'littleleaf linden', 

**Note** using the .tolist() method allows us to output all of the trees in a neighborhood as a list. If we do not use this method, Pandas will print a line for each entry. 

In [ ]:
#get unique species rather than repeats
for item in trees_grouped:
    region = item[0]
    trees = item[1].unique().tolist()
    print(f'{region}:{trees}')

Arverne:['sweetgum', 'Sophora', 'pin oak', 'honeylocust']
Astoria:['London planetree', 'pin oak', 'honeylocust', 'Norway maple', 'Callery pear', 'American linden', 'Sophora', 'tulip-poplar', 'Japanese zelkova', 'black oak', 'littleleaf linden', 'horse chestnut', 'northern red oak', 'red maple', 'Japanese tree lilac', 'hardy rubber tree', 'crab apple', "Schumard's oak", 'cherry', 'sweetgum', 'silver maple']
Bayside:['silver maple', 'Norway maple', 'sweetgum', 'littleleaf linden', 'weeping willow']
Bellerose:['hardy rubber tree']
Bronx:['honeylocust', 'red maple', 'pin oak', 'London planetree', 'American Elm', 'Norway maple', 'ginkgo', 'swamp white oak', 'Sophora', 'Atlantic white cedar', 'white oak', 'Callery pear', 'silver linden', 'Japanese zelkova', 'sawtooth oak', 'Kentucky yellowwood', 'sweetgum', 'Douglas-fir', 'crepe myrtle', 'littleleaf linden', 'American linden', 'eastern redcedar', 'silver maple', 'northern red oak', 'Japanese tree lilac', 'willow oak', 'Siberian elm', 'hedge 

In [ ]:
#get counts for species by neighborhood
for item in trees_grouped:
    region = item[0]
    trees = item[1].unique().tolist()
    counts = item[1].value_counts().tolist()
    print(f'{region}:{trees}:{counts}')

Arverne:['sweetgum', 'Sophora', 'pin oak', 'honeylocust']:[2, 2, 1, 1]
Astoria:['London planetree', 'pin oak', 'honeylocust', 'Norway maple', 'Callery pear', 'American linden', 'Sophora', 'tulip-poplar', 'Japanese zelkova', 'black oak', 'littleleaf linden', 'horse chestnut', 'northern red oak', 'red maple', 'Japanese tree lilac', 'hardy rubber tree', 'crab apple', "Schumard's oak", 'cherry', 'sweetgum', 'silver maple']:[22, 20, 12, 9, 5, 4, 3, 3, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
Bayside:['silver maple', 'Norway maple', 'sweetgum', 'littleleaf linden', 'weeping willow']:[1, 1, 1, 1, 1]
Bellerose:['hardy rubber tree']:[1]
Bronx:['honeylocust', 'red maple', 'pin oak', 'London planetree', 'American Elm', 'Norway maple', 'ginkgo', 'swamp white oak', 'Sophora', 'Atlantic white cedar', 'white oak', 'Callery pear', 'silver linden', 'Japanese zelkova', 'sawtooth oak', 'Kentucky yellowwood', 'sweetgum', 'Douglas-fir', 'crepe myrtle', 'littleleaf linden', 'American linden', 'eastern redceda

In [ ]:
#use .zip() to associate the parallel lists from our analysis into tuples
for item in trees_grouped:
    region = item[0]
    trees = item[1].unique().tolist()
    counts = item[1].value_counts().tolist()
    trees_and_counts = list(zip(trees, counts))

    print(f'{region}:{[tree for tree in trees_and_counts]}')

Arverne:[('sweetgum', 2), ('Sophora', 2), ('pin oak', 1), ('honeylocust', 1)]
Astoria:[('London planetree', 22), ('pin oak', 20), ('honeylocust', 12), ('Norway maple', 9), ('Callery pear', 5), ('American linden', 4), ('Sophora', 3), ('tulip-poplar', 3), ('Japanese zelkova', 2), ('black oak', 2), ('littleleaf linden', 2), ('horse chestnut', 1), ('northern red oak', 1), ('red maple', 1), ('Japanese tree lilac', 1), ('hardy rubber tree', 1), ('crab apple', 1), ("Schumard's oak", 1), ('cherry', 1), ('sweetgum', 1), ('silver maple', 1)]
Bayside:[('silver maple', 1), ('Norway maple', 1), ('sweetgum', 1), ('littleleaf linden', 1), ('weeping willow', 1)]
Bellerose:[('hardy rubber tree', 1)]
Bronx:[('honeylocust', 36), ('red maple', 31), ('pin oak', 15), ('London planetree', 10), ('American Elm', 9), ('Norway maple', 7), ('ginkgo', 7), ('swamp white oak', 4), ('Sophora', 4), ('Atlantic white cedar', 4), ('white oak', 4), ('Callery pear', 3), ('silver linden', 3), ('Japanese zelkova', 2), ('sawt

####Try it yourself:
Use .groupby to bring together the data from two columns. Try to choose a pairing that you think might be interesting and might help us understand the dataset in a different way. Step through the data and print out your results.


**Question:** What if we wanted to save the data we’ve been working with–what might be a good structure to use and what file format would we employ?


In [ ]:
# initialize an empty dictionary for data
trees_dict = {}

#iterate over the dataframe and extract specific information
for item in trees_grouped:
    # region is the neighborhood where the trees are (e.g. Astoria)
    region = item[0]

    # create a unique list of trees to avoid repetitions in species name
    trees = item[1].unique().tolist()

    # count the instances of each species and convert the output to a list
    counts = item[1].value_counts().tolist()
   
    # populate the empty dictionary by setting the first key as region and then creating a dictionary by zipping trees and counts
    trees_dict[region]=dict(zip(trees,counts))


**Notice** that we use the zip function we employed in the previous version of this code but now are wrapping the output in a dictionary by using dict() and assigning that to the key “region”.

In [ ]:
#write the information as a .json() file
import json

trees_path = 'Tree_Species_and_Counts_by_Location.json'

with open(trees_path, 'w') as json_object:
    json.dump(trees_dict, json_object, indent=2)
    print("file written")

file written


In [ ]:
list_o_things = ["bob", 42, ["betty", "fishstick", "bobsled"], "art", 19.5]

for item in list_o_things:
  print(f'{item}')

bob
42
['betty', 'fishstick', 'bobsled']
art
19.5


In [ ]:
for item in list_o_things:
  if isinstance(item, list):
    for thing in item:
      print(f'\t{thing}')
  else:
    print(f'{item}')

bob
42
	betty
	fishstick
	bobsled
art
19.5


In [ ]:
def check_for_list(some_stuff):
  temp_list = []
  for item in some_stuff:
    if isinstance(item, list):
      for thing in item:
        temp_list.append(thing)
    else:
      temp_list.append(item)
  return temp_list

In [ ]:
new_list =check_for_list(list_o_things)
for item in new_list:
  print(item)

bob
42
betty
fishstick
bobsled
art
19.5


In [ ]:
my_stupid_list = ['cat', 'dog', ['GDP', 15, "icecream cone", "doorframe"], 15.6, "woof"]

In [ ]:
second_new_list = check_for_list(my_stupid_list)

for item in second_new_list:
  print(item)

cat
dog
GDP
15
icecream cone
doorframe
15.6
woof


In [ ]:
blank_dict = {}
counter = 1
for item in second_new_list:
  blank_dict[counter] = item
  counter += 1

print(blank_dict)

{1: 'cat', 2: 'dog', 3: 'GDP', 4: 15, 5: 'icecream cone', 6: 'doorframe', 7: 15.6, 8: 'woof'}


```character_dict['full_name'] = full_name(hero_fname[i], hero_lname[i])```

###Pseudo Code
Generate a hero or villain (starting problem)

> requires: parallel lists with:
+ last name
+ first name
+ species
+ characteristics
+ weaknesses

> assemble all elements into a cogent character<br/>
> require a full name from last and first<br/>
> output full character as a string and also as a dictionary<br/>

####Initial solutions
need some way of constructing a full name
> NOTE: could have some emptry strings in first or last, and need to handle that .
> NOTE: could have lists in list indexes, and need to be able to handle that to get all elements
need a way to generate the character so that we get meaningful output.

####Where to start
1. Solve full name.
> creating a function that
>+ takes in first and last
>+ tests for empty string
>+ outputs full name that makes sense

```def full_name(first, last):```<br/>
>```#test first for empty```<br/>
>```#test last for empty```<br/>
>```#return full name``

2. create a "generic" generator that can make a hero or a villain
> What do we need to do here?
>> 1. take in our parallel lists
>> 2. make a full name from the name data in the lists
>>> ex. inside generator make a call to full_name(first, last)
>> 3. test remaining list data for lists
>> 4. capture data (full name, each item from list including items in lists in the list) and write to a dictionary
>> 5. assemble all of the data (full name, any item from the list, including things from lists inside the list) and output to a string using an f string.